### PR TITLE
Adding functionalities from minpy model builder

### PR DESCRIPTION
**Pull request by GaiYu0**
Added functionalities include:

- Binary operators for Layer

- An API enabling user to declare layers via syntax similar to MXNet NDArray operator API

### CODE REVIEW

**Comment by GaiYu0** [comment]
@piiswrong I am not sure why the first test fails. Could you help? Thank you very much!

File: python/mxnet/foo/nn/layer.py [hunk 1]
```diff
@@ -207,6 +222,80 @@ def generic_forward(self, F, x, *args, **kwargs):
         raise NotImplementedError
 
 
+class _Binary(Layer):
```
Comment (piiswrong) [inline python/mxnet/foo/nn/layer.py]: These are very confusing. It's doing branching and merging at the same time. What's the use case? Why not just do it in forward?

File: python/mxnet/foo/nn/legacy.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
+"""
+A legacy neural network package enabling user to
+declare layers via syntax similar to MXNet NDArray API.
+"""
+
+
+import mxnet.ndarray as _nd
```
Comment (piiswrong) [inline python/mxnet/foo/nn/legacy.py]: use relative import ...

File: python/mxnet/foo/nn/legacy.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+Pooling = lambda **kwargs: _LegacyLayer('Pooling', **kwargs)
+RNN = lambda **kwargs: _LegacyLayer('RNN', **kwargs)
+
+# pylint: disable=locally-disabled, invalid-name
+ReLU = lambda: _LegacyLayer('Activation', act_type='relu')
+Sigmoid = lambda: _LegacyLayer('Activation', act_type='sigmoid')
+Tanh = lambda: _LegacyLayer('Activation', act_type='tanh')
+
+# pylint: disable=locally-disabled, invalid-name
+RNNReLU = lambda **kwargs: _LegacyLayer('RNN', mode='rnn_relu', **kwargs)
```
Comment (piiswrong) [inline python/mxnet/foo/nn/legacy.py]: legacy interface should be the same with functional interface. RNN is enough. We don't need these

File: python/mxnet/foo/nn/legacy.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+Deconvolution = lambda **kwargs: _LegacyLayer('Deconvolution', **kwargs)
+Dropout = lambda **kwargs: _LegacyLayer('Dropout', **kwargs)
+Embedding = lambda **kwargs: _LegacyLayer('Embedding', **kwargs)
+FullyConnected = lambda **kwargs: _LegacyLayer('FullyConnected', **kwargs)
+LeakyReLU = lambda **kwargs: _LegacyLayer('LeakyReLU', **kwargs)
+Pooling = lambda **kwargs: _LegacyLayer('Pooling', **kwargs)
+RNN = lambda **kwargs: _LegacyLayer('RNN', **kwargs)
+
+# pylint: disable=locally-disabled, invalid-name
+ReLU = lambda: _LegacyLayer('Activation', act_type='relu')
```
Comment (piiswrong) [inline python/mxnet/foo/nn/legacy.py]: don't need these.

File: python/mxnet/foo/nn/layer.py (hunk 1 above)
Comment (GaiYu0) [inline python/mxnet/foo/nn/layer.py]: Here is an example.

```
net = nn.Sequential()
# add layers ...
net.add(Dense(16) + Activation('relu'))
# add layers ...
```

File: python/mxnet/foo/nn/legacy.py (hunk 2 above)
Comment (GaiYu0) [inline python/mxnet/foo/nn/legacy.py]: It is fixed.

File: python/mxnet/foo/nn/legacy.py (hunk 3 above)
Comment (GaiYu0) [inline python/mxnet/foo/nn/legacy.py]: This has been removed.

File: python/mxnet/foo/nn/legacy.py (hunk 4 above)
Comment (GaiYu0) [inline python/mxnet/foo/nn/legacy.py]: This has been removed.

File: python/mxnet/foo/nn/layer.py (hunk 1 above)
Comment (piiswrong) [inline python/mxnet/foo/nn/layer.py]: I still think this is confusing. Especially when it's overloading + and -.
How about we do it by inheriting layer for now and see if there is a need for this later?

Comment (GaiYu0) [inline python/mxnet/foo/nn/layer.py]: I agree. I have removed operator overloading. Only the explicit form (`Add(lhs, rhs)`) remains now.

**Comment by mli** [comment]
why we need legacy here?

**Comment by GaiYu0** [comment]
Legacy enables users to declare imperative layers via interfaces similar to MXNet NDArray API.

File: python/mxnet/foo/nn/layer.py (hunk 1 above)
Comment (piiswrong) [inline python/mxnet/foo/nn/layer.py]: I don't think Add/sub are good names. I would expect Add to add two NDArrays, not split then merge two layers.

The documentation also needs improvement.

Comment (GaiYu0) [inline python/mxnet/foo/nn/layer.py]: I have removed these layers completely. Although they are useful, they might not be very intuitive for users. I will write a more detailed guide for legacy layers later.

**Comment by piiswrong** [comment]
Please checkout upstream nn again and cherry-pick your commits

**Pull request closed without merging**